src/SSHDebugPS/PipeConnection.cs

```diff
@@ -0,0 +1,251 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Globalization;
+using Microsoft.VisualStudio.Debugger.Interop.UnixPortSupplier;
+
+namespace Microsoft.SSHDebugPS
+{
+    internal abstract class PipeConnection : Connection
+    {
+        private readonly object _lock = new object();
+        private readonly IPipeTransportSettings _settings;
+        private readonly Connection _outerConnection;
+
+        private readonly ShellExecutionManager _shellExecutionManager;
+        private readonly LinkedList<IRawShell> _shellList = new LinkedList<IRawShell>();
+        private bool _isClosed;
+        private string _name;
+
+        public override string Name => _name;
+
+        protected Connection OuterConnection => _outerConnection;
+        protected IPipeTransportSettings TransportSettings => _settings;
+        protected bool IsClosed => _isClosed;
+
+        /// <summary>
+        /// Create a new pipe connection object
+        /// </summary>
+        /// <param name="pipeTransportSettings">Settings</param>
+        /// <param name="outerConnection">[Optional] the SSH connection (or maybe something else in future) used to connect to the target.</param>
+        /// <param name="name">The full name of this connection</param>
+        /// <param name="container">The name of the container. For local, this is the same as 'name'.</param>
+        public PipeConnection(IPipeTransportSettings pipeTransportSettings, Connection outerConnection, string name, int? timeout)
+        {
+            Debug.Assert(pipeTransportSettings != null);
+            Debug.Assert(!string.IsNullOrEmpty(name));
+            //Debug.Assert(!string.IsNullOrEmpty(container));
+
+            _name = name;
+            _settings = pipeTransportSettings;
+            if (outerConnection != null)
+            {
+                _outerConnection = outerConnection;
+                DefaultTimeout = 15000;
```

Comment:
>15000 [](start = 33, length = 5)

Do we need a progress dialog anywhere? 15 seconds is long enough for people to think things are hung.

Reply:
Andrew pointed out that we should allow the user to cancel and set the timeout to something high. I'm looking at making that happen.

---
In reply to: [253705761](https://github.com/Microsoft/MIEngine/pull/831#discussion_r253705761) [](ancestors = 253705761)